### PR TITLE
Make TF pt-tf equivalence test more aggressive

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Make TF pt-tf equivalence test more aggressive.

After a series of fixes done so far, I think it is a good time to include this more aggressive testing in `master` branch.
(Otherwise, the new models added might have undetected issues. For example, the recent `TFConvNextModel` would have `hidden_states` not transposed to match Pytorch version issue - I tested it on my local branch and informed the author to fix it).

There are still 3 categories of PT/TF inconsistency to address, but they are less urgent in my opinion. See below.
Currently, the test makes a few exception to not test these 3 cases (in order to get green test) - I add `TODO` comments in the code.

TF: @Rocketknight1 @gante 
Test: @LysandreJik  @sgugger 

## TODO in separate PRs

- failing due to the difference of large negative values (used for attn mask) between PT/TF:
  - https://circleci.com/api/v1.1/project/github/huggingface/transformers/379126/output/112/0?file=true&allocation-id=621b51ac72134246517cac0c-0-build%2F6E4C1534
  - albert, convbert, speech_to_text, t5, tapas 
- failing due to cache format difference between PT/TF:
  - https://app.circleci.com/pipelines/github/huggingface/transformers/35203/workflows/d0c0fbb4-661b-4a61-96e7-2fd4a3249f44/jobs/379098/parallel-runs/0/steps/0-112
  - gpt2, t5, speech_to_text
- failing due to outputting loss or not between PT/TF:
  - https://app.circleci.com/pipelines/github/huggingface/transformers/35206/workflows/d2533f38-6877-4db5-9eb6-da92e6421e11/jobs/379144/parallel-runs/0/steps/0-112
  - flaubert, funnel, transfo_xl, xlm

### CODE REVIEW

File: tests/test_modeling_tf_common.py [hunk 1]
```diff
@@ -380,65 +366,191 @@ def test_pt_tf_model_equivalence(self):
...
+
+            tf_keys = [k for k, v in tfo.items() if v is not None]
+            pt_keys = [k for k, v in pto.items() if v is not None]
+
+            self.assertEqual(tf_keys, pt_keys)
+            check_outputs(tfo, pto, model_class, names=tf_keys)
+
+            # check the case where `labels` is passed
+            has_labels = any(
+                x in tf_inputs_dict_maybe_with_labels for x in ["labels", "next_sentence_label", "start_positions"]
```
Comment (gante) [inline tests/test_modeling_tf_common.py:475]: We might want to think about a better way to pick up the label name, otherwise this list might get long.

**Review by gante — APPROVED** [review]
thorough tests <3 and +1 for leaving the plan as comments, makes it easier to review.

File: tests/test_modeling_tf_common.py [hunk 2]
```diff
@@ -380,65 +366,191 @@ def test_pt_tf_model_equivalence(self):
...
+                tf_nans = np.copy(np.isnan(tfo))
+                pt_nans = np.copy(np.isnan(pto))
+
+                pto[tf_nans] = 0
+                tfo[tf_nans] = 0
+                pto[pt_nans] = 0
+                tfo[pt_nans] = 0
+
+                max_diff = np.amax(np.abs(tfo - pto))
+                self.assertLessEqual(max_diff, 1e-5)
```
Comment (Rocketknight1) [inline tests/test_modeling_tf_common.py:411]: Will we need some kind of relative tolerance here? 1e-5 is a small allowable difference for potentially large values!

**Review by Rocketknight1 — APPROVED** [review]
Overall, this looks good. If I understand correctly, it makes the following major changes:

1) Much smaller tolerances for differences between PT + TF outputs

    Yes

2) Verifies that all output keys are the same across both models.

    Yes

3) Crossloading is done in-memory instead of saving and loading a checkpoint:

    Yes (it was done this way previously)

Is that correct? Is there any other important parts that I missed?

   It's correct, and no you didn't miss anything. But just a remark: we test 2 cases: `labels` passed to model / `labels` not passed to model.
   (previously, we only test the case without passing `labels`)

File: tests/test_modeling_tf_common.py (hunk 2 above)
Comment (ydshieh) [reply tests/test_modeling_tf_common.py:411]: Probably yes. Currently I have confidence that `1e-5` would work in our current testing configs/models.
The weights are initialized with a std 0.02 -> and with this setting, the output values won't get too large.
(Lysandre told me that once we go for GPU testing and fp16 precision testing, we might need to deal with larger errors).

File: tests/test_modeling_tf_common.py [hunk 3]
```diff
@@ -380,65 +366,191 @@ def test_pt_tf_model_equivalence(self):
...
-            pt_hidden_states[tf_nans] = 0
-            tf_hidden_states[tf_nans] = 0
-            pt_hidden_states[pt_nans] = 0
-            tf_hidden_states[pt_nans] = 0
+            # Output all for aggressive testing
+            config.output_hidden_states = True
+            # Pure convolutional models have no attention
+            # TODO: use a better and general criteria
+            if "TFConvNext" not in model_class.__name__:
+                config.output_attentions = True
```
[the remark this answers is not in the document]
Comment (ydshieh) [reply tests/test_modeling_tf_common.py:422]: - `output_attentions` attribute always exists in the config: it is defined in `class PretrainedConfig(PushToHubMixin)`.
- But it is possible to check in the model input signature. (Do you prefer me to do it in this PR?)

**Comment by ydshieh** [comment]
Just wondering, after being approved by @gante and @Rocketknight1 , am I allowed to merge by myself ... (?)

File: tests/test_modeling_tf_common.py (hunk 2 above)
Comment (LysandreJik) [reply tests/test_modeling_tf_common.py:411]: Yes, would love to know if this passes on GPU as it's typically tougher on small differences.

The point regarding fp16 was specifically regarding bf16: models trained with bf16 typically have much larger logits, so using an absolute difference is unideal; using relative difference should be used.

Here we do the init ourselves, so it doesn't apply.

File: tests/test_modeling_tf_common.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (LysandreJik) [reply tests/test_modeling_tf_common.py:422]: Hmmm yeah indeed this is unideal. I personally think it's fine to leave it like this for now as there's a single architecture, but we'll likely need an identifier (`PreTrainedModel.can_output_attention: bool`?) if this becomes a trend. I think the best flag/property will be easier to identify once we have a few more examples.

File: tests/test_modeling_tf_common.py [hunk 4]
```diff
@@ -380,65 +366,191 @@ def test_pt_tf_model_equivalence(self):
...
+
+            # TODO: remove this block once the large negative value for attention masks is fixed.
```
Comment (LysandreJik) [inline tests/test_modeling_tf_common.py:424]: Would you like to take a stab at that? :) (If yes, it should be in a separate PR, of course)

**Review by LysandreJik — COMMENTED** [review]
Thank you for your time spent making this better! Two questions:
- Does it run on GPU?
- How long does it take to run? Our test suite is already taking a significant time to run, so aiming for common tests that run as fast as possible is important.

There's quite a bit of model-specific logic which I'm not particularly enthusiastic about (pre-training models + convnext), but I understand why it's rigorous to do it like that here.

File: tests/test_modeling_tf_common.py (hunk 4 above)
Comment (ydshieh) [reply tests/test_modeling_tf_common.py:424]: ### 

> Would you like to take a stab at that? :) (If yes, it should be in a separate PR, of course)

Yes, it is on my plans :)

**Comment by ydshieh** [comment]
> Thank you for your time spent making this better! Two questions:
> 
>     * Does it run on GPU?

    For now, I haven't tested it with GPU. I can run it on the office's  GPU machine this week (a good chance to learn how to connect to those machine!)

> 
>     * How long does it take to run? Our test suite is already taking a significant time to run, so aiming for common tests that run as fast as possible is important.

   Let me measure the timing of this test with the current master and with this PR. Will report it.

> 
> 
> There's quite a bit of model-specific logic which I'm not particularly enthusiastic about (pre-training models + convnext), but I understand why it's rigorous to do it like that here.

File: tests/test_modeling_tf_common.py [hunk 5]
```diff
@@ -380,65 +366,191 @@ def test_pt_tf_model_equivalence(self):
                 else:
                     pt_inputs_dict[name] = torch.from_numpy(key.numpy()).to(torch.long)
 
-            with torch.no_grad():
-                pto = pt_model(**pt_inputs_dict)
-            tfo = tf_model(self._prepare_for_class(inputs_dict, model_class), training=False)
+            return pt_inputs_dict
+
+        def check_outputs(tfo, pto, model_class, names):
```
Comment (sgugger) [inline tests/test_modeling_tf_common.py:371]: Nit: would prefer something more explicit here, like `tf_output` and `pt_output` to make the code easier to read.

File: tests/test_modeling_tf_common.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sgugger) [reply tests/test_modeling_tf_common.py:422]: We will have more architectures in the future (ResNet should be coming soon!) so it would make everyone life easier to have this identifier ;-)

File: tests/test_modeling_tf_common.py [hunk 6]
```diff
@@ -380,65 +366,191 @@ def test_pt_tf_model_equivalence(self):
...
+                    attention_mask = tf.ones_like(attention_mask, dtype=tf.int32)
+                    # (make the first sequence with all 0s attention mask -> to demonstrate the issue)
+                    # (this will fail for `TFWav2Vec2Model`)
+                    # attention_mask = tf.concat(
+                    #     [
+                    #         tf.zeros_like(attention_mask[:1], dtype=tf.int32),
+                    #         tf.cast(attention_mask[1:], dtype=tf.int32)
+                    #     ],
+                    #     axis=0
+                    # )
```
Comment (sgugger) [inline tests/test_modeling_tf_common.py:438]: To clean up?

**Review by sgugger — APPROVED** [review]
Thanks for writing those new tests!

File: tests/test_modeling_tf_common.py (hunk 6 above)
Comment (ydshieh) [reply tests/test_modeling_tf_common.py:438]: Will remove it later - this part will be activated and used for a follow PR to ensure everything works well.

(dealing with the `1e-4`, `1e-9`, `1e-30` values for attn mask. Currently, different values across frameworks will introduce larger errors, when a sequence has all 0s attention mask ).

**Comment by ydshieh** [comment]
For the test timing:

| Model    |      Before (s)    | This PR (s) |
|----------|:-------------:|------:|
| BERT | 7.68  | 6.12 |
| GPT2 |    2.64   | 1.78  |
| BART | 4.69 |    3.52 |
| T5 |  2.92 |  1.87 |
| Total |  4m37 |  5m22 |

It's somehow strange that the new test took less time to run for each above model - although it had a longer total run time.
Do you want me to digger deeper for this, @LysandreJik ?
(one )

[Report before this PR](https://app.circleci.com/pipelines/github/huggingface/transformers/35311/workflows/cf12b350-c7ca-4382-82ea-63c418488620/jobs/380708)

[Report after this PR](https://app.circleci.com/pipelines/github/huggingface/transformers/35231/workflows/001c1bc0-f0bf-4bc6-9823-afb54f4e927d/jobs/379503)

**Comment by ydshieh** [comment]
For this testing on GPU (running on a GCP VM with 1 GPU) 
```
== 143 passed, 5 skipped, 16494 deselected, 16 warnings in 352.17s (0:05:52) ===
```

(after an update to make pytorch models/inputs on the correct device)

This looks good, i.e `1e-5` still works on GPU, at least for the current models

However, I would like to make sure again the PyTorch models and inputs DOES run on GPU by stepping into the execution (I don't like to be fooled by potential mistakes). Will confirm once it is done.

**Comment by LysandreJik** [comment]
It's weird that it took more time when you expected it to take less, no? Can you try running the test suite with `--durations=0` to see all the tests and the time it took for them to run?

**Comment by ydshieh** [comment]
@LysandreJik It's weird I can modified you comment! Do you know why? It's a mistake, I didn't intend to do so, sorry.

**Comment by ydshieh** [comment]
@LysandreJik 

I run the previous test and the new test on GCP CPU VM. Here is what I obtained so far.
(There are full output files at the end)

## Observation

- testing is faster in this PR (I can dig deeper, say running `TFBertModelTest::test_pt_tf_model_equivalence` manually)
- the timing (per model testing) recorded with `-n 8` seems very strange, but the final timing looks reasonable. (I don't know why), but it is not very performing (with 8 workers is only 2x faster than only 1 worker)

## Command I use

```
RUN_PT_TF_CROSS_TESTS=1 python -m pytest -n 1 -v tests -k pt_tf_model_equivalence --durations=0 | tee output.txt
```

## Full outputs with some summary

- Before: previous testing method
- After = new testing method done in this PR

----

- 1 worker (-n 1)

  - before this PR: (333.26s)
[test_pt_tf_model_equivalence_cpu_n1_output_commit_84eaa6acf582206dba33135727dc3bfff05a7e9c_before.txt](https://github.com/huggingface/transformers/files/8164548/test_pt_tf_model_equivalence_cpu_n1_output_commit_84eaa6acf582206dba33135727dc3bfff05a7e9c_before.txt)

    - TFBertModelTest::test_pt_tf_model_equivalence: 7.30s
    - TFGPT2ModelTest::test_pt_tf_model_equivalence: 2.72s

  - after this PR: (241.08s)
[test_pt_tf_model_equivalence_cpu_n1_output_commit_cce5b6a6511880b5c6d7dc1b74d327485c595ffc_after.txt](https://github.com/huggingface/transformers/files/8164551/test_pt_tf_model_equivalence_cpu_n1_output_commit_cce5b6a6511880b5c6d7dc1b74d327485c595ffc_after.txt)

    - TFBertModelTest::test_pt_tf_model_equivalence: 3.44s
    - TFGPT2ModelTest::test_pt_tf_model_equivalence: 1.10s

- 8 workers (-n 8)

  - before this PR: (167.67s)
[test_pt_tf_model_equivalence_cpu_n8_output_commit_84eaa6acf582206dba33135727dc3bfff05a7e9c_before.txt](https://github.com/huggingface/transformers/files/8164558/test_pt_tf_model_equivalence_cpu_n8_output_commit_84eaa6acf582206dba33135727dc3bfff05a7e9c_before.txt)


    - TFBertModelTest::test_pt_tf_model_equivalence: 29.80s (very strange values!)
    - TFGPT2ModelTest::test_pt_tf_model_equivalence: 11.22s

  - after this PR: (153.71s)
[test_pt_tf_model_equivalence_cpu_n8_output_commit_cce5b6a6511880b5c6d7dc1b74d327485c595ffc_after.txt](https://github.com/huggingface/transformers/files/8164561/test_pt_tf_model_equivalence_cpu_n8_output_commit_cce5b6a6511880b5c6d7dc1b74d327485c595ffc_after.txt)


    - TFBertModelTest::test_pt_tf_model_equivalence: 26.40ss
    - TFGPT2ModelTest::test_pt_tf_model_equivalence: 6.08s

**Comment by ydshieh** [comment]
> It's weird that it took more time when you expected it to take less, no? Can you try running the test suite with `--durations=0` to see all the tests and the time it took for them to run?

Sure! I actually copied the 2 results from CircleCI report. But I will run it on GCP on CPU to measure the timing.

**Comment by ydshieh** [comment]
After a deeper investigation, I realized that I made a mistake (accidentally removed a test case in the original version).
This is why the observed running time was shorter than before.

In the original test, there were 2 parts:

- https://github.com/huggingface/transformers/blob/2eb7bb15e771f13192968cd4657c78f76b0799fe/tests/test_modeling_tf_common.py#L362
- https://github.com/huggingface/transformers/blob/2eb7bb15e771f13192968cd4657c78f76b0799fe/tests/test_modeling_tf_common.py#L401

The last commit fixed it, and now the correct running time is:

| original test | new test |
| -------------:| -------------:|
| 54.85s  | 66.77s  |

@LysandreJik Let me know if this increasement is acceptable, thanks.

**Comment by ydshieh** [comment]
After a thorough verification and a few fixes, this PR is ready (again) from my side.

I would love sgugger (and Lysandre when he is back) to check it again, and gante & Rocketknight1 if they want (no particular TF-related change compared to the last commit).

The following summary might save some review (again) time

- **Main fixes** (since the last review):

  - In the last commit, the following line forgot to use `to_tuple()`, and the method `check_output` (previous version) only deal with `tuple` or `tensor`. So half test cases passed just because no check was performed at all. (my bad ...).
https://github.com/huggingface/transformers/blob/6f2025054dc49a42dc3de82cbea22f2a5913f122/tests/test_modeling_tf_common.py#L447
This is fixed now, and also `check_output` now raises error if a result is not tested (if its type is not in `[tuple, list, tensor]`).

  - In the last commit, I accidentally deleted the following test case (`check the results after save/load checkpoints`, which exists on the current `master`).
https://github.com/huggingface/transformers/blob/6f2025054dc49a42dc3de82cbea22f2a5913f122/tests/test_modeling_tf_common.py#L576
This is added back now.

  - Make the test also run on GPU (put models/inputs on the correct device)

- **Style changes**

  - Since I have to add back the accidentally deleted test cases , I added `check_pt_tf_models` to avoid duplicated large code block.
  - `tfo` -> `tf_outputs` and `pto` -> `pt_outputs`

I run this new version of test, and with the small tolerance `1e-5`, it pass both on CPU and GPU :tada:!

In order to be super sure, I also ran it on CPU/GPU for 100 times (very aggressive :fire: :fire: )! All models passed for this test, except `TFSpeech2TextModel`. See ...

Regarding the running time, I will measure it and post the results in the next comment.

**Comment by ydshieh** [comment]
In terms of running time:

- Circle CI
  - current : [56.11s](https://circleci.com/api/v1.1/project/github/huggingface/transformers/384755/output/112/0?file=true&allocation-id=62238c1d4f148c1110a1efa8-0-build%2F1610CB69)
  - this PR : [61.21s](https://circleci.com/api/v1.1/project/github/huggingface/transformers/384731/output/112/0?file=true&allocation-id=622385347ccc3f3894ec7697-0-build%2F351FF724)

File: tests/test_modeling_tf_common.py [hunk 7]
```diff
@@ -380,23 +367,214 @@ def test_pt_tf_model_equivalence(self):
...
+            # Allow `list` because `(TF)TransfoXLModelOutput.mems` is a list of tensors.
+            if type(tf_outputs) in [tuple, list]:
+                self.assertEqual(type(tf_outputs), type(pt_outputs))
+                self.assertEqual(len(tf_outputs), len(pt_outputs))
+                if type(names) == tuple:
+                    for tfo, pto, name in zip(tf_outputs, pt_outputs, names):
+                        check_outputs(tfo, pto, model_class, names=name)
+                elif type(names) == str:
+                    for idx, (tfo, pto) in enumerate(zip(tf_outputs, pt_outputs)):
+                        check_outputs(tfo, pto, model_class, names=f"{names}_{idx}")
```
Comment (sgugger) [inline tests/test_modeling_tf_common.py:397]: Still think `pt_output` and `tf_output` would be better names :-)

File: tests/test_modeling_tf_common.py [hunk 8]
```diff
@@ -380,23 +367,214 @@ def test_pt_tf_model_equivalence(self):
...
+                    attention_mask = tf.ones_like(attention_mask, dtype=tf.int32)
+                    # (make the first sequence with all 0s attention mask -> to demonstrate the issue)
+                    # (this will fail for `TFWav2Vec2Model`)
+                    # attention_mask = tf.concat(
+                    #     [
+                    #         tf.zeros_like(attention_mask[:1], dtype=tf.int32),
+                    #         tf.cast(attention_mask[1:], dtype=tf.int32)
+                    #     ],
+                    #     axis=0
+                    # )
```
Comment (sgugger) [inline tests/test_modeling_tf_common.py:559]: Should this be cleaned up then?

**Review by sgugger — APPROVED** [review]
Still good to me! Left two nits.

File: tests/test_modeling_tf_common.py (hunk 8 above)
Comment (ydshieh) [reply tests/test_modeling_tf_common.py:559]: This block should be activated in the future:

  1. we want to clean up (next steps) the large negative values like `1e-9` etc. for attention mask
  2. currently, this block will fail due to the different such values
  3. once 1.) is done, we should enable this block to make sure no regression & the new models work in the same manner as the existing ones
  4. (If you prefer, I can remove this block in this PR, and add it back once we are ready for it)

(I probably said previously in the wrong way - this block is intended to be kept (and enabled) rather than removed, sorry)

**Review by LysandreJik — APPROVED** [review]
Sounds good to me, thank you for working on it @ydshieh!
